jalib/jfilesystem.cpp

```diff
@@ -151,6 +151,46 @@ jalib::Filesystem::DirName(const dmtcp::string &str)
   return str.substr(0, lastSlash);
 }
 
+void jalib::Filesystem::DirName(char *dirname, const char *fname)
+{
+  size_t len = strlen(fname);
+
+  if (strcmp(fname, "/") == 0 || strcmp(fname, ".") == 0 || fname[0] == '\0') {
+    strcpy(dirname, fname);
+    return;
+  }
+
+  if (strcmp(fname, "..") == 0) {
+    strcpy(dirname, ".");
```

Comment:
Is this a bug?  If fname is '..', shouldn't the dirname also be '..'?

Reply:
For '..' it should return '.' according to the manpage. But there was a separate bug and I also followed your suggestion about lastslash.